### ISSUE TITLE
Matplotlib images often do not display on the correct cells

### ISSUE THREAD

**Issue by JoaoFelipe** [jupyter/notebook#5568]
Some cells do not output the matplotlib image as their results. Instead the image appears as an output of the next cell and some images even combine with the output of the next cell. It is also not consistent: some cells do output their images, while others don't.

Example:

In [1]:
```
%matplotlib inline
from matplotlib import pyplot as plt
x = range(15)
y = [n for n in x]
plt.plot(x, y)
```
Out[1]:
```
[<matplotlib.lines.Line2D at 0x7f1d4b12dd60>]
```
(No other display here)

In [2]:
```
plt.plot(x, y)
```
Out[2]:
```
[<matplotlib.lines.Line2D at 0x7f1d4b171e80>]
```
![image](https://user-images.githubusercontent.com/327789/86316169-74937a00-bc02-11ea-8194-581e4327a809.png)

In [3]:
```
plt.plot(x, y)
```

Out[3]:
```
[<matplotlib.lines.Line2D at 0x7f1d4b08a310>]
```
![image](https://user-images.githubusercontent.com/327789/86316209-93920c00-bc02-11ea-851f-83e16f895843.png)

Note that the image of In [1] kept the matplotlib fig for In [2] and it shows an orange line. However, In [3] has its own figure (with the expected blue line that should appear in all three cells)

If I repeat the code, the bug persists:

In [4]:
```
x = range(15)
y = [n for n in x]
plt.plot(x, y)
```
Out[4]:
```
[<matplotlib.lines.Line2D at 0x7f1d4b0616d0>]
```
<No display>

In [5]:
```
print('????')
```
????
![image](https://user-images.githubusercontent.com/327789/86316498-48c4c400-bc03-11ea-85f6-6a4ae11e4114.png)

In [5] should not have this figure. It should appear in the results of In [4]

----------------------

I first noticed the bug running the following versions:
```
ipykernel==5.3.0
ipython==7.16.1
ipython-genutils==0.2.0
jupyter-client==6.1.5
jupyter-core==4.6.3
jupyterlab==2.1.5
jupyterlab-server==1.1.5
notebook==6.0.3
```

But I already tried to clone all these repository and it didn't fix the issue:
```
ipykernel.git@8ec2c4689614342483b7b3d85fe2a8536eef1527
ipython.git@ba11b22ef9f8c94d1ead775dc0ef69043ba9605c
ipython-genutils==0.2.0
jupyter_client.git@288b9d0927bcbf7d3e954c37d14e552f3f373324
jupyter_core.git@ac712354ec2c9377a290acca03a91712f991c889
jupyterlab.git@750771ecda8cb1a800c88047496f72656f225272
jupyterlab-server==1.1.5
notebook@b152dd314decda6edbaee1756bb6f6fc50c50f9f
```

I have an old conda environment that run this notebook properly:
```
ipykernel==5.1.2
ipython==7.8.0
ipython-genutils==0.2.0
jupyter-client==5.3.4
jupyter-core==4.6.0
jupyterlab==1.1.4
jupyterlab-server==1.0.6
notebook==6.0.1
```

Other relevant system information:
```
 'os_name': 'posix'
 'platform': 'Linux-4.15.0-20-generic-x86_64-with-glibc2.10',
 'sys_executable': '/home/joao/anaconda3/envs/raw38/bin/python',
 'sys_platform': 'linux',
 'sys_version': '3.8.3 (default, May 19 2020, 18:47:26) \n[GCC 7.3.0]'}
```

PS: I'm not sure this is the right place. The bug occurs both on `jupyter notebook` and on `jupyter lab`, but it does not seem to occur on `ipython` (which is the package that implements the line magic `%matplotlib inline`, I guess). Of course, `ipython` only display the text/plain representation of the image
```ipython
In [1]: %matplotlib inline 
   ...: from matplotlib import pyplot as plt 
   ...: x = range(15) 
   ...: y = [n for n in x] 
   ...: plt.plot(x, y)                                                          
Out[1]: [<matplotlib.lines.Line2D at 0x7f2febca42b0>]
<Figure size 432x288 with 1 Axes>
```

PPS: While developing a [Jupyter Lab extension](https://github.com/dew-uff/julynter/blob/68170b3/src/linterlab/notebookhandler.ts#L382) (which was disabled during all these tests), I sent a code to execute on the kernel and added a callback for it. The kernel did not reply with the result of my code. Instead it replied with an image:
```TypeScript
// session: ISessionConnection;
const kernel = session.kernel;
const future = kernel.requestExecute('hello', false);
future.onIOPub = (msg: KernelMessage.IIOPubMessage): void => {
  const msgType = msg.header.msg_type;
  if (msgType === 'execute_result' || msgType ===  'display_data') {
    const payload = msg.response as IExecuteResult;
    console.log(payload.data['text/plain']);  // Usually it displays the result of my code ('hello'). But during the bug, it displays "<Figure size 432x288 with 1 Axes>"
  }
};
```

I suspect that it is not recognizing the correct cell to update with the image output

**Comment by JoaoFelipe** [comment]
It seems related to #5532

**Comment by JoaoFelipe** [comment]
Update: I installed the versions of the packages that I mentioned before "as working" in a Python 3.8 environment, and the bug also occurs.
Then, I installed the most recent version of Jupyter in a Python 3.7 environment and everything works. 

Also, if I surround the list comprehension into a lambda, the code works as expected:
In [1]:
```
%matplotlib inline
from matplotlib import pyplot as plt
x = range(15)
y = (lambda: [n for n in x])()
plt.plot(x, y)
```
![image](https://user-images.githubusercontent.com/327789/86424383-cac2f480-bcb8-11ea-9654-fa8715376c3b.png)

**Comment by JoaoFelipe** [comment]
Fixed with ipykernel 5.3.1
